server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1302,78 +1398,125 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer,
      * 5. Deadlock.  T2 is waiting for T1's shared lock to be released but
      *    T1 is waiting for T2's shared lock to be released.
      *
-     * The solution is to create a longer transaction and grab an exclusive lock
+     * The solution was to create a longer transaction and grab an exclusive lock
      * on the cp_consumer row (using a select for update) at the start of the transaction.
      * The other thread will then wait for the exclusive lock to be released instead of
      * deadlocking.
      *
-     * See BZ #1274074
+     * Another effort on the solution removed compliance status evaluation from
+     * this thread and created a separate asynchronous job to accomplish that,
+     * thus removing the need to hold that exclusive lock on cp_consumer
+     *
+     * See BZ #1274074 and git history for details
      */
     @Transactional
-    protected Entitlement addOrUpdateEntitlement(Consumer consumer, Pool pool,
-        Entitlement entitlement, Integer quantity, boolean generateUeberCert,
-        CallerType caller)
+    protected List<Entitlement> addOrUpdateEntitlements(Consumer consumer,
+            Map<String, Integer> poolQuantityMap,
+            Map<String, Entitlement> entitlements, boolean generateUeberCert, CallerType caller)
         throws EntitlementRefusedException {
         // Because there are several paths to this one place where entitlements
         // are granted, we cannot be positive the caller obtained a lock on the
-        // pool
-        // when it was read. As such we're going to reload it with a lock
+        // pool when it was read. As such we're going to reload it with a lock
         // before starting this process.
-        log.info("Locking pool: {}", pool.getId());
-        pool = poolCurator.lockAndLoad(pool);
 
-        if (quantity > 0) {
+        log.debug("Locking pools: {}", poolQuantityMap.keySet());
+
+        List<Pool> pools = poolCurator.lockAndLoad(poolQuantityMap.keySet());
+
+        for (Pool pool : pools) {
+            log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
+        }
```

Comment:
Should we check if we are in debug mode before iterating the pools to log a message?